### ISSUE TITLE
Hot Reload not working when Typescript Function is in subdirectory

### ISSUE THREAD

**Issue by panmona** [Azure/azure-functions-host#7795]
The hot reloading is not working when a Typescript function is in a subdirectory and you're using one of the following ways to start it:
```
func start --prefix {dir}
func start--script-root {dir}
cd {dir} && func start
```

# Repro
I made a reproduction repository with which I can reliably reproduce this issue: https://github.com/panmau/azure-function-hot-reload-repro

- Get the function running locally
- `curl http://localhost:7071/api/Test`
- Observe the logged output
- Make a change, for example: Hello World to Hello Universe
- Rerun `curl http://localhost:7071/api/Test`
- The logging output doesn't change! The compiled js file did. The function doesn't seem to restart after the recompilation.

The repro is based on the normal configuration that you receive with:
```
func init
func new 
```
The difference is that I moved the directory of the function one level down, added one of the ways to start the function to package.json and updated the @azure/function package.

# Environment
```
func --version
3.0.3568

OS
Manjaro Linux 21.0.7 Ornara
```

Let me know whether you need any more info and if you can reproduce it.
I am willing to contribute the bugfix but I need some guidance in which area it could possibly be.

**Comment by ejizba** [comment]
@panmau yes the issue is with `watchDirectories`. It turns out the host watches the root of the app (see [here](https://github.com/Azure/azure-functions-host/blob/c833bfe3ba76dd83bb7cf01febae29c09ab7e8af/src/WebJobs.Script.WebHost/FileMonitoringService.cs#L157)) and only uses `watchDirectories` to _filter_ file change events (see [here](https://github.com/Azure/azure-functions-host/blob/c833bfe3ba76dd83bb7cf01febae29c09ab7e8af/src/WebJobs.Script.WebHost/FileMonitoringService.cs#L281)). This would need to be fixed by the host - it would have to actually watch those directories instead of just use them to filter. I'm not entirely sure of the implications of that change, though. Transferring to host repo

This issue is also related to https://github.com/Azure/azure-functions-host/issues/5373 - we plan to support more complex file structures in general.

**Comment by satvu** [comment]
This is a confirmed **host bug** (not Core Tools). The root cause is in `FileMonitoringService.cs`:

- The `FileWatcherEventSource` only watches `RootScriptPath` and its subdirectories.
- `watchDirectories` in `host.json` is only used as a **filter** on events already detected within that root — it does not create additional watchers for directories outside the root.
- If a TypeScript project compiles to a directory outside the script root, changes are never detected.

This remains relevant for local development scenarios with TypeScript (and potentially other compiled languages) where the output directory may live outside the script root.

**Next steps:**

1. Verify whether this is still reproducible with the **Node.js v4 programming model** (which has built-in TypeScript support and a different project structure). If v4 handles this differently, the issue may only affect legacy v3 model projects.
2. If still relevant: `FileMonitoringService.InitializeFileWatchers()` should create additional `FileWatcherEventSource` instances for any `watchDirectories` entries that resolve outside `RootScriptPath`.
3. Alternatively, document that `watchDirectories` only works for paths *within* the script root, and recommend structuring projects so the output directory is nested under it.

---
*This response was generated with AI assistance and reviewed by an engineer.*
